frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn active_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		T::Staking::stake(&who).map(|s| s.active).unwrap_or_default()
+	}
+
+	pub(crate) fn to_vote(balance: BalanceOf<T>) -> VoteWeight {
+		let total_issuance = T::Currency::total_issuance();
+		<T::Staking as StakingInterface>::CurrencyToVote::to_vote(balance, total_issuance)
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn on_stake_update(who: &T::AccountId, _: Option<Stake<T::AccountId, BalanceOf<T>>>) {
+		if let Ok(current_stake) = T::Staking::stake(who) {
+			let current_active = current_stake.active;
+
+			// if this is a nominator
+			if let Some(_) = T::Staking::nominations(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active));
+			}
+
+			if T::Staking::is_validator(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active));
+			}
+		}
+	}
+
+	fn on_nominator_update(who: &T::AccountId, _prev_nominations: Vec<T::AccountId>) {
+		// NOTE: We ignore the result here, because this method can be called when the nominator
+		// is already in the list, just changing their nominations.
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(Self::active_stake_of(who)));
+	}
+
+	fn on_validator_update(who: &T::AccountId) {
+		let self_stake = Self::active_stake_of(who);
+		// maybe update sorted list.
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(self_stake));
+	}
+
+	fn on_validator_remove(who: &T::AccountId) {
+		let _ = T::VoterList::on_remove(who);
+	}
+
+	fn on_nominator_remove(who: &T::AccountId, _nominations: Vec<T::AccountId>) {
+		let _ = T::VoterList::on_remove(who);
+	}
+
+	fn on_unstake(_who: &T::AccountId) {}
+}
+
+/// A wrapper for a given `SortedListProvider` that disables insert/update/remove operations,
+/// effectively rendering it read-only, except for unsafe operations.
+pub struct TrackedList<T, S, P>(sp_std::marker::PhantomData<(T, S, P)>);
+
+impl<T: Config, S: Bounded + Saturating + Zero, P: SortedListProvider<T::AccountId, Score = S>>
+	SortedListProvider<T::AccountId> for TrackedList<T, S, P>
+{
+	type Error = P::Error;
+	type Score = P::Score;
+	fn iter() -> Box<dyn Iterator<Item = T::AccountId>> {
+		P::iter()
+	}
+
+	fn iter_from(
+		start: &T::AccountId,
+	) -> Result<Box<dyn Iterator<Item = T::AccountId>>, Self::Error> {
+		P::iter_from(start)
+	}
+
+	fn count() -> u32 {
+		P::count()
+	}
+
+	fn contains(id: &T::AccountId) -> bool {
+		P::contains(id)
+	}
+
+	fn get_score(id: &T::AccountId) -> Result<Self::Score, Self::Error> {
+		P::get_score(id)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn try_state() -> Result<(), &'static str> {
+		P::try_state()
+	}
+
+	fn on_insert(_id: T::AccountId, _score: Self::Score) -> Result<(), Self::Error> {
+		unreachable!()
```

Comment:
This is a really bad idea, as if staking accidentally call into these, it will cause a panic. You could leave the implementation empty, but that would lead to an inconsistent state.. 

I see what you are tying to do, it if there were no issues with feature flags the `Sorted` sub-trait is the way to go, but at this point the downsides outweight the positives.

Reply:
Assuming we are fully covered with tests - panics should not happen. 

What is the alternative? Having empty method bodies is probably worse, because it does not pop up in tests.

Reply:
Yeah, all the options are kinda horrible, ReadOnly abstraction was good, but kinda ugly.

The only alternative I can think of is introducing some defensive checks in this abstraction, while still delegating to the original list. That way we get the best of two worlds - easy spot in testing and no potential panics in the runtime.

Reply:
defensive is good 👍